lib-ui-compose/src/commonMain/kotlin/com/ustadmobile/libuicompose/view/person/edit/PersonEditScreen.kt

```diff
@@ -221,16 +226,25 @@ fun PersonEditScreen(
 
         if (uiState.usernameVisible){
             OutlinedTextField(
-                modifier = Modifier.testTag("username").fillMaxWidth().defaultItemPadding(),
+                modifier = Modifier.testTag("username").fillMaxWidth().defaultItemPadding().onKeyEvent { keyEvent ->
+                    if (keyEvent.type == KeyEventType.KeyDown) {
+                        FilterUsernameUseCase.shouldBlockKeyEvent(keyEvent.utf16CodePoint.toChar())
+                    } else false
+                },
                 value = uiState.person?.username ?: "",
                 label = { Text(stringResource(MR.strings.username)) },
                 enabled = uiState.fieldsEnabled,
                 isError = uiState.usernameError != null,
                 singleLine = true,
-                onValueChange = {
-                    onPersonChanged(uiState.person?.shallowCopy{
-                        username = it
-                    })
+                onValueChange = { newValue ->
+                    val lastChar = newValue.lastOrNull()
+                    if (lastChar == null || !FilterUsernameUseCase.shouldBlockKeyEvent(lastChar)) {
```

Comment:
I already pointed out on https://github.com/UstadMobile/UstadMobile/pull/986 - "This logic fails when anytime the user has changed any character which is not at the end of the text field." - comment on LoginViewModel.kt

Please remove all value change event processing based on the flawed assumption that the last character in the string is the one that was just changed.